modules/space/horizonsfile.cpp

```diff
@@ -0,0 +1,877 @@
+        // In case of comets it says the Source in the top
+        foundTarget = true;
+    }
+
+    ResultCode result = ResultCode::UnknownError;
+    while (fileStream.good() && line.find("$$SOE") == std::string::npos) {
+        // Selected time range too big and step size too small?
+        if (line.find("change step-size") != std::string::npos) {
+            fileStream.close();
+            return ResultCode::ErrorSize;
+        }
+
+        // Selected time range too big for avalable time span?
+        if (line.find("STEP_SIZE too big") != std::string::npos) {
+            fileStream.close();
+            return ResultCode::ErrorSpan;
+        }
+
+        // Outside valid time range?
+        if (line.find("No ephemeris for target") != std::string::npos) {
+            // Available time range is located several lines before this in the file
+            // The avalable time range is persed later
+            fileStream.close();
+            return ResultCode::ErrorTimeRange;
+        }
+
+        // Valid Observer?
+        if (line.find("No site matches") != std::string::npos ||
+            line.find("Cannot find central body") != std::string::npos)
+        {
+            fileStream.close();
+            return ResultCode::ErrorNoObserver;
+        }
+
+        // Are observer and target the same?
+        if (line.find("disallowed") != std::string::npos)
+        {
+            fileStream.close();
+            return ResultCode::ErrorObserverTargetSame;
+        }
+
+        // Enough data?
+        if (line.find("Insufficient ephemeris data") != std::string::npos)
+        {
+            fileStream.close();
+            return ResultCode::ErrorNoData;
+        }
+
+        // Multiple Observer stations?
+        if (line.find("Multiple matching stations found") != std::string::npos) {
+            result = ResultCode::MultipleObserverStations;
+        }
+
+        // Multiple matching major bodies?
+        if (line.find("Multiple major-bodies match string") != std::string::npos) {
+            // Target
+            if (!foundTarget) {
+                // If target was not found then it is the target that has multiple matches
+                result = ResultCode::MultipleTarget;
+            }
+            // Observer
+            else {
+                result = ResultCode::MultipleObserver;
+            }
+        }
+
+        // Multiple matching small bodies?
+        if (line.find("Small-body Index Search Results") != std::string::npos) {
+            // Small bodies can only be targets not observers
+            result = ResultCode::MultipleTarget;
+        }
+
+        // No Target?
+        if (line.find("No matches found") != std::string::npos) {
+            fileStream.close();
+            return ResultCode::ErrorNoTarget;
+        }
+
+        std::getline(fileStream, line);
+    }
+
+    if (result != ResultCode::UnknownError) {
+        fileStream.close();
```

Comment:
since the fileStream is closed in all the checks below, maybe it can just be done before this if-statement ?

Reply:
If it is closed here, then the check for `good()` a few lines down is invalid. We need to check if we found the start of data before we reached the end of the file i.e. the file needs to still be `good()` for it to be valid in the end of this function.